core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java

```diff
@@ -190,34 +191,75 @@ protected void performClear() {
       super.performClear();
    }
 
+   /**
+    * This method will make sure that if there are more elements than there should be that it will remove one at
+    * a time until it is finally below or equal to the configured threshold.
+    *
+    * This is done by first acquiring the lock on the lru node to first check if the size has been violated or not.
+    * Then while holding the lock if the size is too large it will:
+    * <ul>
+    *    <li>Retrieve the hashCode of the key for the head entry</li>
+    *    <li>With that hashCode try to acquire the write lock for the given key</li>
+    *    <li>If the lock is acquired the lru lock can be released, the entry is removed, and the write lock is unlocked</li>
+    * </ul>
+    * If the lock was not able to be acquired for the given key we have to try a slightly different order of handling.
+    * <ul>
+    *    <li>First we unlock the lru node lock</li>
+    *    <li>Then we wait on the key write lock until it can be acquired</li>
+    *    <li>Once the write lock is acquired we reacquire the lru node lock (this order is required to prevent deadlock)</li>
+    *    <li>We check the size again just to make sure we should attempt to remove an entry</li>
+    *    <li>Lastly we check the hashCode of the head pointer again make sure it shares the same lock
+    *        (we don't care if the entry changed as long as it shares the same lock)</li>
+    *    <li>If lock is the same we unlock lru node and then finally remove the entry and release write lock</li>
+    * </ul>
+    * This entire process is repeated again until the current size is smaller or equal to the maximum size.
```

Comment:
I was thinking of something more high-level, like this:

    * This method repeatedly removes the head of the LRU list until there the current size is less than or equal to `maxSize`.
    *
    * We need to hold the LRU lock in order to check the current size and to read the head entry,
    * and then we need to hold the head entry's write lock in order to remove it.
    * The problem is that the correct acquisition order is entry write lock first, LRU lock second,
    * and we need to hold the LRU lock so that we know which entry write lock to acquire.
    *
    * To work around it, we first try to acquire the entry write lock without blocking.
    * If that fails, we release the LRU lock and we acquire the locks in the correct order, hoping that
    * the LRU head doesn't change while we wait. Because the entry write locks are striped, we actually
    * tolerate a LRU head change as long as the new head entry is in the same lock stripe.
    * If the LRU list head changes, we release both locks and try again.

Writing this got me thinking that we could improve the algorithm a little: if the current head entry has a different lock stripe at the end, we can skip the release and re-acquisition of the LRU lock for the next iteration.

Reply:
Yes, that's what I meant. And you're right, tracking whether `lruLock` is acquired adds quite a bit of complexity, so let's leave it out.